File: src/incidents/list.js
/**
 * Lambda function: Listar incidentes con filtros
 * Endpoint: GET /incidents
 * Requiere: Autenticación
 */
import { queryItems, scanItems, batchGetItems } from '../utils/dynamodb.js';
import { requireAuth, isStudent, isWorker } from '../utils/auth.js';
import { successResponse, unauthorized, internalError } from '../utils/responses.js';

const INCIDENTS_TABLE = process.env.INCIDENTS_TABLE;
const USERS_TABLE = process.env.USERS_TABLE;

/**
 * Lista incidentes con filtros avanzados
 * 
 * Query Parameters:
 * - status: pending, assigned, in_progress, resolved, closed
 * - priority: low, medium, high, urgent
 * - category: electricidad, plomeria, etc.
 * - assignedTo: userId del trabajador
 * - building: nombre del edificio
 * - limit: número de resultados (default: 50, max: 100)
 * - lastKey: para paginación
 */
export const handler = async (event) => {
  try {
    console.log('=== LISTAR INCIDENTES ===');
    
    // Verificar autenticación
    let user;
    try {
      user = requireAuth(event);
      console.log('Usuario autenticado:', user.userId, user.role);
    } catch (error) {
      console.error('Error de autenticación:', error.message);
      return unauthorized(error.message);
    }

    // Obtener query parameters
    const params = event.queryStringParameters || {};
    console.log('Query params:', params);
    
    const status = params.status;
    const priority = params.priority;
    const category = params.category;
    let assignedTo = params.assignedTo;
    const building = params.building;
    let limit = parseInt(params.limit || '50');
    const lastKeyStr = params.lastKey;

    // Validar límite
    if (limit > 100) {
      limit = 100;
    }

    // Parsear lastKey si existe
    let lastKey = null;
    if (lastKeyStr) {
      try {
        lastKey = JSON.parse(lastKeyStr);
      } catch (error) {
        console.error('Error parsing lastKey:', error);
      }
    }

    // Todos los usuarios pueden ver todos los incidentes con filtros
    let incidentsResult;
    
    console.log('Obteniendo todos los incidentes con filtros para usuario:', user.role);
    incidentsResult = await getFilteredIncidents(
      status,
      priority,
      category,
      assignedTo,
      building,
      limit,
      lastKey
    );

    console.log(`Incidentes encontrados: ${incidentsResult.items.length}`);

    // Obtener información de usuarios (reportedBy, assignedTo)
    const incidents = incidentsResult.items;
    const enrichedIncidents = await enrichIncidentsWithUsers(incidents);

    // Preparar respuesta
    const responseData = {
      incidents: enrichedIncidents,
      count: enrichedIncidents.length,
      limit
    };

    if (incidentsResult.lastEvaluatedKey) {
      responseData.lastEvaluatedKey = JSON.stringify(
        incidentsResult.lastEvaluatedKey
      );
    }

    console.log('=== LISTADO EXITOSO ===');
    return successResponse(
      `${enrichedIncidents.length} incidente(s) encontrado(s)`,
      responseData
    );

  } catch (error) {
    console.error('=== ERROR EN LIST INCIDENTS ===');
    console.error('Tipo:', error.name);
    console.error('Mensaje:', error.message);
    console.error('Stack:', error.stack);
    return internalError('Error al listar incidentes');
  }
};

/**
 * Obtiene incidentes reportados por un estudiante
 */
async function getStudentIncidents(studentId, status, limit, lastKey) {
  const options = {
    limit,
    lastEvaluatedKey: lastKey
  };

  // Construir filter expression
  if (status) {
    options.filterExpression = 'reportedBy = :studentId AND #status = :status';
    options.expressionValues = { 
      ':studentId': studentId,
      ':status': status 
    };
    options.expressionNames = { '#status': 'status' };
  } else {
    options.filterExpression = 'reportedBy = :studentId';
    options.expressionValues = { ':studentId': studentId };
  }

  return scanItems(INCIDENTS_TABLE, options);
}

/**
 * Obtiene incidentes para un trabajador (asignados + pendientes)
 */
async function getWorkerIncidents(
  workerId,
  status,
  priority,
  category,
  building,
  limit,
  lastKey
) {
  let filterExpression = '';
  const expressionValues = {};
  const expressionNames = { '#status': 'status' };

  // Si hay filtro de status específico
  if (status) {
    if (status === 'pending') {
      filterExpression = '#status = :status';
      expressionValues[':status'] = 'pending';
    } else {
      filterExpression = 'assignedTo = :workerId AND #status = :status';
      expressionValues[':workerId'] = workerId;
      expressionValues[':status'] = status;
    }
  } else {
    // Mostrar incidentes asignados a él O pendientes
    filterExpression = '(assignedTo = :workerId) OR (#status = :pending)';
    expressionValues[':workerId'] = workerId;
    expressionValues[':pending'] = 'pending';
  }

  // Agregar filtros adicionales
  if (priority) {
    filterExpression += ' AND priority = :priority';
    expressionValues[':priority'] = priority;
  }
  if (category) {
    filterExpression += ' AND category = :category';
    expressionValues[':category'] = category;
  }
  if (building) {
    filterExpression += ' AND #location.#building = :building';
    expressionValues[':building'] = building;
    expressionNames['#location'] = 'location';
    expressionNames['#building'] = 'building';
  }

  return scanItems(INCIDENTS_TABLE, {
    filterExpression,
    expressionValues,
    expressionNames,
    limit,
    lastEvaluatedKey: lastKey
  });
}

/**
 * Obtiene incidentes con filtros (admin)
 */
async function getFilteredIncidents(
  status,
  priority,
  category,
  assignedTo,
  building,
  limit,
  lastKey
) {
  // Si hay status, usar el índice StatusCreatedAtIndex
  if (status) {
    const result = await queryItems(
      INCIDENTS_TABLE,
      '#status = :status',
      { ':status': status },
      {
        indexName: 'StatusCreatedAtIndex',
        limit,
        scanForward: false,
        lastEvaluatedKey: lastKey,
        expressionNames: { '#status': 'status' }
      }
    );

    // Aplicar filtros adicionales en memoria
    let items = result.items;
    if (priority) {
      items = items.filter(i => i.priority === priority);
    }
    if (category) {
      items = items.filter(i => i.category === category);
    }
    if (assignedTo) {
      items = items.filter(i => i.assignedTo === assignedTo);
    }
    if (building) {
      items = items.filter(i => i.location?.building === building);
    }

    return { ...result, items };
  }

  // Si hay priority, usar el índice PriorityIndex
  if (priority) {
    const result = await queryItems(
      INCIDENTS_TABLE,
      'priority = :priority',
      { ':priority': priority },
      {
        indexName: 'PriorityIndex',
        limit,
        scanForward: false,
        lastEvaluatedKey: lastKey
      }
    );

    // Aplicar filtros adicionales
    let items = result.items;
    if (category) {
      items = items.filter(i => i.category === category);
    }
    if (building) {
      items = items.filter(i => i.location?.building === building);
    }

    return { ...result, items };
  }

  // Si solo hay filtros simples, hacer scan
  let filterExpression = null;
  const expressionValues = {};
  const expressionNames = {};

  if (category) {
    filterExpression = 'category = :category';
    expressionValues[':category'] = category;
  }
  if (building) {
    const buildingFilter = '#location.#building = :building';
    expressionValues[':building'] = building;
    expressionNames['#location'] = 'location';
    expressionNames['#building'] = 'building';
    filterExpression = filterExpression 
      ? `${filterExpression} AND ${buildingFilter}`
      : buildingFilter;
  }

  // Si no hay filtros, devolver todos los incidentes
  if (!filterExpression) {
    return scanItems(INCIDENTS_TABLE, {
      limit,
      lastEvaluatedKey: lastKey
    });
  }

  return scanItems(INCIDENTS_TABLE, {
    filterExpression,
    expressionValues,
    expressionNames: Object.keys(expressionNames).length > 0 ? expressionNames : undefined,
    limit,
    lastEvaluatedKey: lastKey
  });
}

/**
 * Agrega información de usuarios (reportedBy, assignedTo) a los incidentes
 */
async function enrichIncidentsWithUsers(incidents) {
  if (!incidents || incidents.length === 0) {
    return [];
  }

  // Recolectar todos los user IDs únicos
  const userIds = new Set();
  for (const incident of incidents) {
    if (incident.reportedBy) {
      userIds.add(incident.reportedBy);
    }
    if (incident.assignedTo) {
      userIds.add(incident.assignedTo);
    }
  }

  // Obtener usuarios en batch
  const users = {};
  if (userIds.size > 0) {
    const keys = Array.from(userIds).map(uid => ({ userId: uid }));
    try {
      const userItems = await batchGetItems(USERS_TABLE, keys);
      for (const user of userItems) {
        users[user.userId] = user;
      }
    } catch (error) {
      console.error('Error obteniendo usuarios:', error);
      // Continuar sin enriquecer si hay error
    }
  }

  // Enriquecer incidentes
  const enriched = [];
  for (const incident of incidents) {
    const enrichedIncident = { ...incident };

    // Agregar info del reportador
    const reportedById = incident.reportedBy;
    if (reportedById && users[reportedById]) {
      const reporter = users[reportedById];
      enrichedIncident.reportedBy = {
        userId: reporter.userId,
        name: reporter.name,
        email: reporter.email,
        studentCode: reporter.studentCode || null
      };
    } else if (reportedById) {
      // Si no se encontró el usuario, mantener solo el ID
      enrichedIncident.reportedBy = {
        userId: reportedById,
        name: 'Usuario desconocido',
        email: null
      };
    }

    // Agregar info del trabajador asignado
    const assignedToId = incident.assignedTo;
    if (assignedToId && users[assignedToId]) {
      const worker = users[assignedToId];
      enrichedIncident.assignedTo = {
        userId: worker.userId,
        name: worker.name,
        email: worker.email,
        specialty: worker.specialty || 'General',
        workloadPoints: worker.workloadPoints || 0
      };
    } else if (assignedToId) {
      enrichedIncident.assignedTo = {
        userId: assignedToId,
        name: 'Trabajador desconocido',
        email: null
      };
    } else {
      // Si no hay assignedTo, asegurarse de que sea null en la respuesta
      enrichedIncident.assignedTo = null;
    }

    enriched.push(enrichedIncident);
  }

  return enriched;
}

File: src/incidents/list.py
"""
Lambda function: Listar incidentes con filtros
Endpoint: GET /incidents
Requiere: Autenticación
"""
import json
import os
from typing import Dict, Any, Optional

# Importar utilidades
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from utils.responses import success_response, bad_request, unauthorized, internal_error
from utils.dynamodb import (
    query_items, scan_items, get_user_by_id, batch_get_items,
    INCIDENTS_TABLE, decimal_to_float
)
from utils.auth import require_auth, is_student, is_worker
from boto3.dynamodb.conditions import Key, Attr


def handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lista incidentes con filtros avanzados
    
    Query Parameters:
    - status: pending, assigned, in_progress, resolved, closed
    - priority: low, medium, high, urgent
    - category: electricidad, plomeria, etc.
    - assignedTo: userId del trabajador
    - building: nombre del edificio
    - limit: número de resultados (default: 50, max: 100)
    - lastKey: para paginación
    
    Examples:
    GET /incidents?status=pending
    GET /incidents?priority=urgent&building=Pabellón A
    GET /incidents?assignedTo=usr_worker123
    """
    try:
        # Verificar autenticación
        try:
            user = require_auth(event)
        except ValueError as e:
            return unauthorized(str(e))
        
        # Obtener query parameters
        params = event.get('queryStringParameters') or {}
        
        status = params.get('status')
        priority = params.get('priority')
        category = params.get('category')
        assigned_to = params.get('assignedTo')
        building = params.get('building')
        limit = int(params.get('limit', 50))
        last_key_str = params.get('lastKey')
        
        # Validar límite
        if limit > 100:
            limit = 100
        
        # Parsear lastKey si existe
        last_key = None
        if last_key_str:
            try:
                last_key = json.loads(last_key_str)
            except:
                pass
        
        # Determinar qué incidentes puede ver el usuario
        incidents_result = None
        
        # Si es estudiante, solo ve sus propios incidentes
        if is_student(user):
            incidents_result = _get_student_incidents(
                user['userId'], 
                status, 
                limit, 
                last_key
            )
        
        # Si es trabajador, ve incidentes asignados a él + pendientes
        elif is_worker(user):
            if assigned_to and assigned_to != user['userId']:
                # Workers no pueden ver incidentes de otros workers
                assigned_to = user['userId']
            
            incidents_result = _get_worker_incidents(
                user['userId'],
                status,
                priority,
                category,
                building,
                limit,
                last_key
            )
        
        # Si es admin, ve todos los incidentes con filtros
        else:
            incidents_result = _get_filtered_incidents(
                status,
                priority,
                category,
                assigned_to,
                building,
                limit,
                last_key
            )
        
        # Obtener información de usuarios (reportedBy, assignedTo)
        incidents = incidents_result['items']
        enriched_incidents = _enrich_incidents_with_users(incidents)
        
        # Convertir Decimals a float/int
        enriched_incidents = decimal_to_float(enriched_incidents)
        
        # Preparar respuesta
        response_data = {
            'incidents': enriched_incidents,
            'count': len(enriched_incidents),
            'limit': limit
        }
        
        if incidents_result.get('lastEvaluatedKey'):
            response_data['lastEvaluatedKey'] = json.dumps(
                incidents_result['lastEvaluatedKey']
            )
        
        return success_response(
            message=f"{len(enriched_incidents)} incidente(s) encontrado(s)",
            data=response_data
        )
    
    except Exception as e:
        print(f"Error en list incidents: {str(e)}")
        import traceback
        traceback.print_exc()
        return internal_error(f"Error al listar incidentes: {str(e)}")


def _get_student_incidents(
    student_id: str,
    status: Optional[str],
    limit: int,
    last_key: Optional[Dict]
) -> Dict[str, Any]:
    """Obtiene incidentes reportados por un estudiante"""
    filter_expr = Attr('reportedBy').eq(student_id)
    
    if status:
        filter_expr = filter_expr & Attr('status').eq(status)
    
    return scan_items(
        INCIDENTS_TABLE,
        filter_expression=filter_expr,
        limit=limit,
        last_evaluated_key=last_key
    )


def _get_worker_incidents(
    worker_id: str,
    status: Optional[str],
    priority: Optional[str],
    category: Optional[str],
    building: Optional[str],
    limit: int,
    last_key: Optional[Dict]
) -> Dict[str, Any]:
    """Obtiene incidentes para un trabajador (asignados + pendientes)"""
    
    # Si hay filtro de status específico
    if status:
        if status == 'pending':
            # Mostrar pendientes (para que pueda tomarlos)
            filter_expr = Attr('status').eq('pending')
        else:
            # Mostrar sus incidentes con ese status
            filter_expr = Attr('assignedTo').eq(worker_id) & Attr('status').eq(status)
    else:
        # Mostrar incidentes asignados a él O pendientes
        filter_expr = Attr('assignedTo').eq(worker_id) | Attr('status').eq('pending')
    
    # Agregar filtros adicionales
    if priority:
        filter_expr = filter_expr & Attr('priority').eq(priority)
    if category:
        filter_expr = filter_expr & Attr('category').eq(category)
    if building:
        filter_expr = filter_expr & Attr('location.building').eq(building)
    
    return scan_items(
        INCIDENTS_TABLE,
        filter_expression=filter_expr,
        limit=limit,
        last_evaluated_key=last_key
    )


def _get_filtered_incidents(
    status: Optional[str],
    priority: Optional[str],
    category: Optional[str],
    assigned_to: Optional[str],
    building: Optional[str],
    limit: int,
    last_key: Optional[Dict]
) -> Dict[str, Any]:
    """Obtiene incidentes con filtros (admin)"""
    
    # Si hay status, usar el índice StatusCreatedAtIndex
    if status:
        result = query_items(
            INCIDENTS_TABLE,
            key_condition=Key('status').eq(status),
            index_name='StatusCreatedAtIndex',
            limit=limit,
            scan_forward=False,
            last_evaluated_key=last_key
        )
        
        # Aplicar filtros adicionales
        items = result['items']
        if priority:
            items = [i for i in items if i.get('priority') == priority]
        if category:
            items = [i for i in items if i.get('category') == category]
        if assigned_to:
            items = [i for i in items if i.get('assignedTo') == assigned_to]
        if building:
            items = [i for i in items if i.get('location', {}).get('building') == building]
        
        result['items'] = items
        return result
    
    # Si hay assignedTo, usar el índice AssignedToIndex
    elif assigned_to:
        result = query_items(
            INCIDENTS_TABLE,
            key_condition=Key('assignedTo').eq(assigned_to),
            index_name='AssignedToIndex',
            limit=limit,
            scan_forward=False,
            last_evaluated_key=last_key
        )
        
        # Aplicar filtros adicionales
        items = result['items']
        if priority:
            items = [i for i in items if i.get('priority') == priority]
        if category:
            items = [i for i in items if i.get('category') == category]
        if building:
            items = [i for i in items if i.get('location', {}).get('building') == building]
        
        result['items'] = items
        return result
    
    # Si hay priority, usar el índice PriorityIndex
    elif priority:
        result = query_items(
            INCIDENTS_TABLE,
            key_condition=Key('priority').eq(priority),
            index_name='PriorityIndex',
            limit=limit,
            scan_forward=False,
            last_evaluated_key=last_key
        )
        
        # Aplicar filtros adicionales
        items = result['items']
        if category:
            items = [i for i in items if i.get('category') == category]
        if building:
            items = [i for i in items if i.get('location', {}).get('building') == building]
        
        result['items'] = items
        return result
    
    # Si solo hay filtros simples, hacer scan
    else:
        filter_expr = None
        
        if category:
            filter_expr = Attr('category').eq(category)
        if building:
            building_filter = Attr('location.building').eq(building)
            filter_expr = building_filter if not filter_expr else filter_expr & building_filter
        
        return scan_items(
            INCIDENTS_TABLE,
            filter_expression=filter_expr,
            limit=limit,
            last_evaluated_key=last_key
        )


def _enrich_incidents_with_users(incidents: list) -> list:
    """Agrega información de usuarios (reportedBy, assignedTo) a los incidentes"""
    
    if not incidents:
        return []
    
    # Recolectar todos los user IDs únicos
    user_ids = set()
    for incident in incidents:
        if incident.get('reportedBy'):
            user_ids.add(incident['reportedBy'])
        if incident.get('assignedTo'):
            user_ids.add(incident['assignedTo'])
    
    # Obtener usuarios en batch
    users = {}
    if user_ids:
        keys = [{'userId': uid} for uid in user_ids]
        from utils.dynamodb import USERS_TABLE
        user_items = batch_get_items(USERS_TABLE, keys)
        users = {u['userId']: u for u in user_items}
    
    # Enriquecer incidentes
    enriched = []
    for incident in incidents:
        enriched_incident = incident.copy()
        
        # Agregar info del reportador
        reported_by_id = incident.get('reportedBy')
        if reported_by_id and reported_by_id in users:
            reporter = users[reported_by_id]
            enriched_incident['reportedBy'] = {
                'userId': reporter['userId'],
                'name': reporter.get('name'),
                'email': reporter.get('email'),
                'studentCode': reporter.get('studentCode')
            }
        
        # Agregar info del trabajador asignado
        assigned_to_id = incident.get('assignedTo')
        if assigned_to_id and assigned_to_id in users:
            worker = users[assigned_to_id]
            enriched_incident['assignedTo'] = {
                'userId': worker['userId'],
                'name': worker.get('name'),
                'email': worker.get('email'),
                'specialty': worker.get('specialty'),
                'workloadPoints': worker.get('workloadPoints', 0)
            }
        
        enriched.append(enriched_incident)
    
    return enriched